src/decopatch/main.pyi

```diff
@@ -0,0 +1,72 @@
+from typing import Any, Callable, Optional, Protocol, TypeVar, overload
+
+from decopatch.utils_disambiguation import FirstArgDisambiguation
+from decopatch.utils_modes import SignatureInfo
+
+try:
+    from typing import ParamSpec
+except ImportError:
+    from typing_extensions import ParamSpec
+
+P = ParamSpec("P")
+F = TypeVar("F", bound=Callable[..., Any])
+
+class _Decorator(Protocol[P]):
+    """
+    This is callable Protocol, to distinguish between cases where
+    created decorator is called as `@decorator` or `@decorator()`
+    """
+
+    # decorator is called without parenthesis: @decorator
+    @overload
+    def __call__(self, func: F) -> F: ...
+    # decorator is called with options or parenthesis: @decorator(some_option=...)
+    @overload
+    def __call__(self, *args: P.args, **kwargs: P.kwargs) -> Callable[[F], F]: ...
+
+# @function_decorator is called without options or parenthesis
+@overload
+def function_decorator(
+    enable_stack_introspection: Callable[P, Any],
+    custom_disambiguator: Callable[[Any], FirstArgDisambiguation] = ...,
+    flat_mode_decorated_name: Optional[str] = ...,
+) -> _Decorator[P]: ...
+
+# @function_decorator() is called with options or parenthesis.
+@overload
+def function_decorator(
+    enable_stack_introspection: bool = ...,
+    custom_disambiguator: Callable[[Any], FirstArgDisambiguation] = ...,
+    flat_mode_decorated_name: Optional[str] = ...,
+) -> Callable[[Callable[P, Any]], _Decorator[P]]: ...
+def class_decorator(
+    enable_stack_introspection: bool = ...,
+    custom_disambiguator: Callable[[Any], FirstArgDisambiguation] = ...,
+    flat_mode_decorated_name: Optional[str] = ...,
+): ...
+def decorator(
+    is_function_decorator: bool = ...,
+    is_class_decorator: bool = ...,
+    enable_stack_introspection: bool = ...,
+    custom_disambiguator: Callable[[Any], FirstArgDisambiguation] = ...,
+    use_signature_trick: bool = ...,
+    flat_mode_decorated_name: str = ...,
+): ...
+def create_decorator(
+    impl_function,
+    is_function_decorator: bool = ...,
+    is_class_decorator: bool = ...,
+    enable_stack_introspection: bool = ...,
+    custom_disambiguator: Callable[[Any], FirstArgDisambiguation] = ...,
+    use_signature_trick: bool = ...,
+    flat_mode_decorated_name: Optional[str] = ...,
+): ...
+def create_no_args_decorator(
+    decorator_function, function_for_metadata: Any | None = ...
+): ...
+def create_kwonly_decorator(
+    sig_info: SignatureInfo, decorator_function, disambiguator, function_for_metadata
+): ...
+def create_general_case_decorator(
+    sig_info: SignatureInfo, impl_function, disambiguator, function_for_metadata
+): ...
```

Comment:
I am not sure that the pyi file needs to be complete (if no definition is found, I guess mypy will use the type comments in the function itself). If you confirm, these can be commented out safely.

Reply:
Unfortunately, if pyi file exists type-checkers will use it instead of py, so it must be complete.

In theory, `stubtest` from mypy could be used to verify that pyi matches py, but that would require some effort, i haven't been able to run it yet.

```
> stubtest decopatch.main
error: not checking stubs due to mypy build errors:
src/decopatch/utils_modes.py:7: error: Skipping analyzing "funcsigs": module is installed, but missing library stubs orpy.typed marker
src/decopatch/utils_modes.py:7: error: Name "signature" already defined (possibly by an import)
src/decopatch/utils_modes.py:7: error: Name "Parameter" already defined (possibly by an import)
src/decopatch/utils_modes.py:200: error: Name "Signature" is not defined
src/decopatch/utils_modes.py:370: error: Name "Signature" is not defined
src/decopatch/utils_calls.py:13: error: Name "DecoratorUsageInfo" is not defined
src/decopatch/__init__.py:12: error: Skipping analyzing "setuptools_scm": module is installed, but missing library stubs or py.typed marker
src/decopatch/__init__.py:12: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
```

Reply:
ok. Let's leave it like this then